src/amsterdam_schema/publish/cli.py

```diff
@@ -86,15 +86,43 @@ def fetch_local_as_publishable(
     return publishable_paths
 
 
-def get_index_file_obj(publishable_paths: List[List[str]]) -> BytesIO:
-    index = {}
-    for path_parts in publishable_paths:
+def get_index_file_obj(publishable_paths: List[List[str]], files_root: Path) -> BytesIO:
+    """Create index file of all the datasets in JSON format.
+
+    For every dataset.json file it will add an entry of the form:
+    `dataset_id:'path/to/dataset'`.
+    When a versions.json file is encountered, its path will be included
+    and any dataset.json files with the same id in subdirectories ignored.
+    """
+    index: Dict[str, str] = {}
+    versions_file = "versions.json"
+    dataset_file = "dataset.json"
+    for path_parts in sorted(publishable_paths, key=len):
         if path_parts[1] != "datasets":
             continue
-        dataset_ext = path_parts[-1]
+        if path_parts[-1] not in (versions_file, dataset_file):
+            continue
+
+        # Grab id from dataset file
+        with open(str(files_root / "/".join(path_parts))) as f:
+            key = json.load(f)["id"]
+
         folder = "/".join(path_parts[2:-1])
-        dataset = splitext(dataset_ext)[0]
-        index[dataset] = f"{folder}/{dataset}"
+
+        # Check for duplicate ids
+        if key in index:
+            # throw exception if this is not a dataset in a subfolder of existing entry
+            if path_parts[-1] == versions_file or not folder.startswith(index[key]):
+                raise Exception(
+                    f"Datasets with duplicate id:'{key}' \
+                    in {folder} and {index[key]}."
+                )
+
+            # Ignore datasets in subfolders of a versions.json
+            if path_parts[-1] != versions_file:
+                continue
+
+        index[key] = folder
     return BytesIO(json.dumps(index).encode("utf-8"))
```

Comment:
Not part of your change, but `utf-8` is the default encoding, hence this can be simplified to:

```python
return BytesIO(json.dumps(index).encode())
```

Reply:
Explicit is better than implicit, though ;)

Reply:
Surely you are not advocating to extend each call to `open` with `"r"` as the second argument? ;-)

What I am saying is that defaults should be relied upon. It is what they are meant for.

Reply:
Fair enough.